daemon/containerd/image_push.go

```diff
@@ -53,6 +53,10 @@ func (i *ImageService) PushImage(ctx context.Context, sourceRef reference.Named,
 				return err
 			}
 
+			if len(imgs) == 0 {
+				return fmt.Errorf("An image does not exist locally with the tag: %s", reference.FamiliarName(sourceRef))
+			}
```

Comment:
Not related to this PR, but we should look at the order of events here; we currently (both in the snapshotter and graphdriver variants) start sending a progress message;

```go
progress.Messagef(out, "", "The push refers to repository [%s]", sourceRef.Name())
```

And we do so _BEFORE_ we perform this check; this means that we already have sent a `200` status header;

```bash
curl -v -XPOST -H 'X-Registry-Auth: e30K' --unix-socket /var/run/docker.sock http://localhost/images/nosuchimage/push
*   Trying /var/run/docker.sock...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /images/nosuchimage/push HTTP/1.1
> Host: localhost
> User-Agent: curl/7.58.0
> Accept: */*
> X-Registry-Auth: e30K
>
< HTTP/1.1 200 OK
< Api-Version: 1.43
< Content-Type: application/json
< Docker-Experimental: false
< Ostype: linux
< Server: Docker/24.0.2 (linux)
< Date: Mon, 30 Oct 2023 19:00:15 GMT
< Transfer-Encoding: chunked
<
{"status":"The push refers to repository [docker.io/library/nosuchimage]"}
{"errorDetail":{"message":"An image does not exist locally with the tag: nosuchimage"},"error":"An image does not exist locally with the tag: nosuchimage"}
* Connection #0 to host localhost left intact
```

While there will still be many possible other scenarios where we may fail during this action, I think this is one we can catch before we start the actual push; similar to how we handle a missing auth header;

```bash
curl -v -XPOST --unix-socket /var/run/docker.sock http://localhost/images/nosuchimage/push
* Trying /var/run/docker.sock...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /images/nosuchimage/push HTTP/1.1
> Host: localhost
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< Api-Version: 1.43
< Content-Type: application/json
< Docker-Experimental: false
< Ostype: linux
< Server: Docker/24.0.2 (linux)
< Date: Mon, 30 Oct 2023 19:03:50 GMT
< Content-Length: 94
<
{"message":"bad parameters and missing X-Registry-Auth: invalid X-Registry-Auth header: EOF"}
* Connection #0 to host localhost left intact
```

If we do, we should probably

- Make sure this error is a `InvalidParameter` (at least, I think that's the correct status to use)
- And while we're at it, we should consider fixing the error message to not be capitalised (`An image` -> `an image`); perhaps cleanup the error overall, because it's a bit awkwardly phrased.